src/coprocessor/gadgets.rs

```diff
@@ -213,29 +537,247 @@ mod test {
         let a_one = g.get_allocated_ptr(z_one.tag(), *z_one.value()).unwrap();
 
         // proper list
-        let a_list = construct_list(
-            &mut cs.namespace(|| "proper list"),
-            &g,
-            &store,
-            &[&a_one, &a_one],
-            None,
-        )
-        .unwrap();
+        let a_list = construct_list(&mut cs, &g, &store, &[&a_one, &a_one], None).unwrap();
         let z_list = store.hash_ptr(&store.list(vec![one, one]));
         assert_eq!(a_list.tag().get_value(), Some(z_list.tag_field()));
         assert_eq!(a_list.hash().get_value(), Some(*z_list.value()));
 
         // improper list
-        let a_list = construct_list(
-            &mut cs.namespace(|| "improper list"),
+        let a_list =
+            construct_list(&mut cs, &g, &store, &[&a_one, &a_one], Some(a_one.clone())).unwrap();
+        let z_list = store.hash_ptr(&store.improper_list(vec![one, one], one));
+        assert_eq!(a_list.tag().get_value(), Some(z_list.tag_field()));
+        assert_eq!(a_list.hash().get_value(), Some(*z_list.value()));
+    }
```

Comment:
I know this PR isn't the one that introduced the pattern, but still: 
```rust
fn value_as_zptr(a: &AllocatedPtr<Fq>) -> Option<ZPtr<Fq>> {
        a.tag()
            .get_value()
            .and_then(|t| Tag::from_field(&t))
            .and_then(|tag| a.hash().get_value().map(|hash| ZPtr::from_parts(tag, hash)))
    }
```
Then: 
```sh
comby -matcher .rs -extensions .rs -in-place -jobs 10 -exclude-dir "target,.github" -stats -timeout 15 'assert_eq!(:[a].tag().get_value(), Some(:[b].tag_field())); assert_eq!(:[a].hash().get_value(), Some(*:[b].value()));' 'assert_eq!(value_as_zptr(&:[a]), Some(:[b]));'
```
with https://comby.dev/ will get you a much nicer test file.